Update navigator.canShare, navigator.share, Web Share API

hamishwillee opened the pull request:
Further updates to `navigator.canShare`.

The main reason is that following some digging, I have verified that the message is a bit flawed. Essentially it can return invalid `true` if it does not understand data. It is however useful for feature detection.
I added a simple example using pseudo code for that, and explained it. 

Also removed the non-standard header and experimental flag. IMO it is not to either - though I am testing that in BCD. BCD also has spec update. 

Also noted that the value for the data is a "base set". This can be extended. Lots of work to be done in Web Share API doc (still to be created).

hamishwillee committed:
Further clarifications on navigator.canShare

hamishwillee committed:
Apply suggestions from code review

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee commented on files/en-us/web/api/navigator/canshare/index.html:
```diff
@@ -9,44 +9,72 @@
+  <p><strong>Warning:</strong> The method is flawed in that it may return <code>true</code> for invalid share data that specifies more than one property:</p>
+  <ul>
+    <li>Unsupported fields will be ignored: if there are other supported and valid fields in the share data parameter the method will return <code>true</code> even though part of the data is not shareable.</li>
+    <li>The specification does not require that new supported fields in the data are valid. It is possible that extension fields might be dropped even if valid, and the data still be considered "sharable".</li>
```
There isn't one. I'm removing this future looking stuff based on Joe's comments.

hamishwillee commented on files/en-us/web/api/navigator/canshare/index.html:
```diff
@@ -9,44 +9,72 @@
-<p>The <strong><code>Navigator.canShare()</code></strong> method of the Web Share API returns <code>true</code> if a call to {{domxref("navigator.share()")}} would succeed.</p>
+<p>The <strong><code>Navigator.canShare()</code></strong> method of the <a href="/en-US/docs/Web/API/Web_Share_API">Web Share API</a> returns <code>true</code> if the equivalent call to {{domxref("navigator.share()")}} would succeed.</p>
+
+<p>The method can be used to feature-test if a data types is supported on a particular platform, including both the existing fields and new share data type extensions.</p>
```
Fair enough. Removed

hamishwillee commented on files/en-us/web/api/navigator/canshare/index.html:
```diff
@@ -9,44 +9,72 @@
   The <strong><code>canShare()</code></strong> method will return <code>false</code> if the permission is supported on the platform but has not been granted.</p>
 
+<div class="notecard warning">
+  <p><strong>Warning:</strong> The method is flawed in that it may return <code>true</code> for invalid share data that specifies more than one property:</p>
```
Thanks. I will fix.

hamishwillee commented on files/en-us/web/api/navigator/canshare/index.html:
```diff
@@ -9,44 +9,72 @@
+      <li><code>files</code>: An array of files to be shared.</li>
+    </ul>
+   <p>Additional options may be supported in future.
+     Properties that are not supported on a particular platform will be ignored (the method will return <code>false</code> if there are no other valid properties in the data).</p>
```
@jpmedley Thanks. You're right. I will replace the grumpy subjective viewpoint with a simple statement about the comparison only being on "data properties understood by the implementation".

hamishwillee committed:
Make more objective

hamishwillee commented on files/en-us/web/api/navigator/canshare/index.html:
```diff
@@ -9,13 +9,17 @@
+<p>The Web Share API is gated by the <a href="/en-US/docs/Web/HTTP/Headers/Feature-Policy/web-share">web-share</a> permission policy.
   The <strong><code>canShare()</code></strong> method will return <code>false</code> if the permission is supported on the platform but has not been granted.</p>
 
+<p>The method will generally also return <code>false</code> if the implementation does not support sharing of an indicated data type, for example files, or if sharing that data type would be considered a "hostile share" by the user-agent.
```
@saschanaz 
- This attempts to capture some general guidance about what might cause `false` to be returned. 
- Is this a reasonable statement? I know that it is true for `files` - but is it the intent of the API?
- When you refer to "implementation" in the spec about files perhaps not being shareable, do you mean "the user agent" or the underlying OS to which the share data is being passed?
- The term "hostile share" is used in the spec. I can imagine what it would mean, but does not appear to be defined. Is there one that I am missing?

saschanaz replied:
The statement is now changed so I'll only answer the last two ones.

>When you refer to "implementation" in the spec about files perhaps not being shareable, do you mean "the user agent" or the underlying OS to which the share data is being passed?

Kinda both, the OS or the user agent not supporting the OS feature properly.

>The term "hostile share" is used in the spec. I can imagine what it would mean, but does not appear to be defined. Is there one that I am missing?

That's from https://github.com/w3c/web-share/issues/127 but admittedly it's way too vague. Filed https://github.com/w3c/web-share/issues/217

hamishwillee commented:
@saschanaz @jpmedley Thanks for the review - adjusted to address your points. A few specifics:
- Removed the info about dependency on Permission API. There is no `web-share` permission - just  a web share "permission policy" permission.
- Does not talk about future extensions, though it does still mention base set of options.
- I now use the terminology  valid/invalid property indicate whether a property is understood by the implementation and "supported" to mean understood and will be evaluated, but not necessarily allowed. 
- I added a clarification of what is likely to return false in https://github.com/mdn/content/pull/8381#discussion_r698115221

hamishwillee committed:
Be more precise about what causes a false validation

hamishwillee committed:
Add navigator.share docs

hamishwillee committed:
Add webshare API overview page

jpmedley commented on files/en-us/web/api/navigator/share/index.html:
```diff
@@ -11,73 +11,117 @@
-  action (usually the user has chosen an application to share to). It will reject
-  immediately if the <var>data</var> parameter is not correctly specified, and will also
-  reject if the user cancels sharing.</p>
+<p>A {{jsxref("Promise")}}. This will be resolved with <code>undefined</code> once the data has been sent to the share target, or rejected with one of the <a href="#exceptions">Exceptions</a> given below.</p>
```
```suggestion
<p>A {{jsxref("Promise")}} that resolves with <code>undefined</code> once the data has been sent to the share target, or rejected with one of the <a href="#exceptions">Exceptions</a> given below.</p>
```

jpmedley commented on files/en-us/web/api/navigator/share/index.html:
```diff
@@ -11,73 +11,117 @@
+  <dt><code>TypeError</code></dt>
+  <dd>The specified share data cannot be validated. Possible reasons include:
+    <ul>
+      <li>The <code>data</code> parameter has been omitted completely or only only contains "unknown" properties. Note that any properties that are not recognized by the user agent are ignored.</li>
```
```suggestion
      <li>The <code>data</code> parameter was omitted completely or only contains properties with unknown values. Note that any properties that are not recognized by the user agent are ignored.</li>
```

jpmedley commented on files/en-us/web/api/navigator/share/index.html:
```diff
@@ -11,73 +11,117 @@
+  <dd>The specified share data cannot be validated. Possible reasons include:
+    <ul>
+      <li>The <code>data</code> parameter has been omitted completely or only only contains "unknown" properties. Note that any properties that are not recognized by the user agent are ignored.</li>
+      <li>URL data is badly formatted.</li>
```
```suggestion
      <li>A URL is badly formatted.</li>
```

jpmedley commented on files/en-us/web/api/navigator/share/index.html:
```diff
@@ -11,73 +11,117 @@
-    href="https://github.com/mdn/dom-examples/blob/master/web-share/index.html">see the
-    source code</a>) there is a button which, when clicked, invokes the Web Share API to
-  share MDN's URL. The JavaScript looks like this:</p>
+<p>The example below shows how a button click can be used to invoke the Web Share API to share MDN's URL.
```
```suggestion
<p>The example below shows a button click invoking the Web Share API to share MDN's URL.
```

jpmedley commented on files/en-us/web/api/navigator/share/index.html:
```diff
@@ -11,73 +11,117 @@
-  data object passed to <code>canShare()</code> only supports the <code>files</code>
-  property. Image, video, audio, and text files can be shared.</p>
+<div class="notecard note">
+  <p><strong>Note:</strong> This code sample handles feature detection by testing for <code>navigator.canShare()</code> rather than for <code>navigator.share()</code>.
```
```suggestion
  <p><strong>Note:</strong> This sample feature detects by testing for <code>navigator.canShare()</code> rather than for <code>navigator.share()</code>.
```

jpmedley commented on files/en-us/web/api/navigator/share/index.html:
```diff
@@ -103,4 +147,6 @@ <h2 id="See_also">See also</h2>
 
 <ul>
   <li>{{domxref("navigator.canShare()")}}</li>
+  <li><a href="https://wpt.live/web-share/">https://wpt.live/web-share/</a> (web platform tests)</li>
```
Why include web platform tests? We don't do that on other web features.

hamishwillee replied:
@jpmedley Mostly because I find them very useful, in particular for testing whether features are _now_ supported on a particular browser despite what the browser-compat-data says. I assume that others might also find them useful, and not necessarily be aware of them. 

As you say, it is inconsistent. I'll remove them.

jpmedley commented on files/en-us/web/api/web_share_api/index.html:
```diff
@@ -0,0 +1,72 @@
+---
+<div>{{DefaultAPISidebar("Web Share API")}}</div>
+
+<p>The <strong>Web Share API</strong> provides a mechanism for sharing text, links, files and other content to an arbitrary <em>share target</em> selected by the user.</p>
```
```suggestion
<p>The <strong>Web Share API</strong> provides a mechanism for sharing text, links, files, and other content to an arbitrary <em>share target</em> selected by the user.</p>
```

jpmedley commented on files/en-us/web/api/web_share_api/index.html:
```diff
@@ -0,0 +1,72 @@
+
+<p>The available share targets are provided using the native mechanisms of the underlying operating system. They might, for example, include the system clipboard or other services, Bluetooth or WiFi, email, contacts or messaging applications, and websites.</p>
+
+<p>The API consists of just two methods:</p>
```
Please format this like the other API overview pages. [Here's an example](https://developer.mozilla.org/en-US/docs/Web/API/Contact_Picker_API#interfaces).

hamishwillee replied:
Thanks. I copied the Battery Status API - in retrospect, perhaps not a good choice.

jpmedley commented on files/en-us/web/api/web_share_api/index.html:
```diff
@@ -0,0 +1,72 @@
+<p>The API consists of just two methods:</p>
+
+<ul>
+  <li>{{domxref("navigator.canShare()")}}: Validate that the specified data is sharable (check if a call with the same data to {{domxref("navigator.share()")}} would succeed).</li>
```
```suggestion
  <li>{{domxref("navigator.canShare()")}}: Returns a boolean indicating whether the specified data sharable; indicates whether a call with the same data to {{domxref("navigator.share()")}} would succeed.</li>
```

hamishwillee replied:
@jpmedley I'm proposing and going to accept thismodification of your wording to remove the text after the semicolon. The method does check if the data is sharable, but there are other reasons why `navigator.share()` might fail.
```suggestion
  <li>{{domxref("navigator.canShare()")}}: Returns a boolean indicating whether the specified data sharable.</li>
```

We could keep the sentence and clarify that, but I think the first part captures the concept sufficiently.

jpmedley commented on files/en-us/web/api/web_share_api/index.html:
```diff
@@ -0,0 +1,72 @@
+
+<ul>
+  <li>{{domxref("navigator.canShare()")}}: Validate that the specified data is sharable (check if a call with the same data to {{domxref("navigator.share()")}} would succeed).</li>
+  <li>{{domxref("navigator.share()")}}: Share the specified data. This method must be called on a button click or other user activation (requires {{Glossary("transient activation")}}).</li>
```
```suggestion
  <li>{{domxref("navigator.share()")}}: Returns a {{jsxref("Promise")}} that resolves if the passed data was successfully shared. This method must be called on a button click or other user activation (requires {{Glossary("transient activation")}}).</li>
```

hamishwillee replied:
@jpmedley Re "resolves if the passed data was successfully shared". Technically if "successfully passed to a share target", since it can fail before actually being shared outside of this API. I have modified but let me know if you think the distinction is not useful. 

```suggestion
  <li>{{domxref("navigator.share()")}}: Returns a {{jsxref("Promise")}} that resolves if the passed data was successfully sent to a share target. This method must be called on a button click or other user activation (requires {{Glossary("transient activation")}}).</li>
```

jpmedley commented on files/en-us/web/api/web_share_api/index.html:
```diff
@@ -0,0 +1,72 @@
+</ul>
+
+<p>The Web Share API is gated by the <a href="/en-US/docs/Web/HTTP/Headers/Feature-Policy/web-share">web-share</a> permission policy.
+  If the policy is supported but has not been granted, both methods will immediately indicate that the data is not sharable.</p>
```
```suggestion
  If the policy is supported but has not been granted, both methods will indicate that the data is not sharable.</p>
```

jpmedley commented on files/en-us/web/api/web_share_api/index.html:
```diff
@@ -0,0 +1,72 @@
+
+<h2 id="Example">Example</h2>
+
+<p>The code fragment below shows how you can share a link using {{domxref("navigator.share()")}} triggered off a button click.</p>
```
```suggestion
<p>The code below shows how you can share a link using {{domxref("navigator.share()")}}, triggered off a button click.</p>
```

hamishwillee committed:
Update files/en-us/web/api/navigator/canshare/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/navigator/canshare/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/navigator/canshare/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/navigator/canshare/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/web_share_api/index.html

hamishwillee committed:
Update files/en-us/web/api/web_share_api/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/web_share_api/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/navigator/canshare/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/web_share_api/index.html

hamishwillee commented on files/en-us/web/api/navigator/canshare/index.html:
```diff
@@ -9,12 +9,21 @@
+  <li>The <code>data</code> parameter has been omitted or only only contains "unknown" properties. Note that any properties that are not recognized by the user agent are ignored.</li>
+  <li>URL data is badly formatted.</li>
+  <li>Files are specified but the implementation does not support file sharing.</li>
+  <li>Sharing the specified data would be considered a "hostile share" by the user-agent.</li>
```
No. I raised that and it is being addressed in https://github.com/w3c/web-share/issues/217. We will have to fix as a post-process.

hamishwillee committed:
Update files/en-us/web/api/navigator/canshare/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/navigator/canshare/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/navigator/share/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/navigator/share/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/navigator/share/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/navigator/share/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/navigator/share/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/navigator/share/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/navigator/share/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/navigator/share/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Update files/en-us/web/api/web_share_api/index.html

Co-authored-by: Joe Medley <jmedley@google.com>

hamishwillee committed:
Make API match the normal format

hamishwillee committed:
Add Landing tag

hamishwillee commented:
Thanks for review @jpmedley. Since then ...
- I accepted all your suggestions (with some minor qualifications). Github wouldn't let me batch merge them - sorry if you got multiple notifications.
- Updated the top level API landing page to match the 'standard'
- Removed the link to web tests

jpmedley commented:
Thanks for your patience.

hamishwillee commented:
Thanks for yours @jpmedley - much appreciated.
